core/pow/engine.go

```diff
@@ -561,3 +576,97 @@ func (e *Engine) BlockData() (uint64, string) {
 	}
 	return e.currentBlock, e.blockHash[e.currentBlock%ringSize]
 }
+
+func getParamsForBlock(block uint64, activeParams []*params) *params {
+	stateInd := 0
+	for i, p := range activeParams {
+		if block >= p.fromBlock && (p.untilBlock == nil || *p.untilBlock >= block) {
+			stateInd = i
+			break
+		}
+	}
+
+	params := activeParams[stateInd]
+	return params
+}
+
+func (e *Engine) GetSpamStatistics(partyID string) *protoapi.PoWStatistic {
+	e.lock.RLock()
+	defer e.lock.RUnlock()
+
+	stats := make([]*protoapi.PoWBlockState, 0, len(e.activeStates)) //
```

Comment:
I don't think this allocation makes sense. There's no real connection between the number of blocks in scope that the party can have transactions on and the number of active states which is the number of network param changes that are in scope.